src/main/kotlin/nl/tudelft/ewi/se/ciselab/testgenie/services/TestCaseDisplayService.kt

```diff
@@ -103,6 +117,85 @@ class TestCaseDisplayService(private val project: Project) {
+            val checkbox = it.first
+            val testName = it.second
+            checkbox.addActionListener {
+                addCheckboxListener(checkbox, testName, report)(it)
+            }
+        }
+    }
+
+    /**
+     * Adds listener to checkbox
+     * If a test is selected, the lines it covers will be highlighted. Otherwise they will not.
+     * Works by replicating the original file within the Workspace, then changing its values and placing the new version in Workspace
+     * @param checkbox the checkbox to add listener to
+     * @param testName the name of the test to add/remove from unhighlightList
+     * @param reportFetched the compactReport with the tests
+     */
+    private fun addCheckboxListener(checkbox: JCheckBox, testName: String, reportFetched: CompactReport): (ActionEvent) -> Unit = {
+        if (checkbox.isSelected) {
+            unhighlightList.remove(testName)
+        } else {
+            unhighlightList.add(testName)
+        }
+
+        val originalResult = project.service<Workspace>().getTestGenerationResult(reportFetched)
+
+        var report = reportFetched
+
+        // Always use the root report. That's the report with all the tests still in it
+        // Otherwise, we will lose information if we deselect multiple tests
+        if (originalCompactReports.containsKey(originalResult.first)) {
+            // Deep copy required
+            report = deepCopy(originalCompactReports.get(originalResult.first) ?: reportFetched)
+        } else {
+            originalCompactReports.put(originalResult.first, deepCopy(reportFetched))
+        }
+
+        // copy report
+        val reportNew: CompactReport = report
+
+        val removedNames = mutableListOf<Pair<String, CompactTestCase?>>()
+
+        // remove tests which are not selected
+        unhighlightList.forEach {
+            val test = reportNew.testCaseList.remove(it)
+            removedNames.add(Pair(it, test))
+        }
+        val coveredLinesNew: MutableSet<Int> = mutableSetOf()
+
+        // only count the covered lines of selected tests
+        report.testCaseList.forEach {
+            coveredLinesNew.addAll(it.value.coveredLines)
+        }
+
+        reportNew.allCoveredLines = coveredLinesNew
+
+        // replace original file in workplace
+        project.service<Workspace>().addPendingResult(originalResult.first, originalResult.second)
+        project.service<Workspace>().receiveGenerationResult(originalResult.first, reportNew)
+    }
+
+    /**
+     * Makes a deep copy of the compact report
+     * Credits to https://medium.com/@gugabernardo/here-is-the-easiest-way-to-deep-copy-an-object-in-kotlin-cf200c2130b
+     * @param report the compactReport to make the copy of
+     * @return copy of compactReport
+     */
+    private fun deepCopy(report: CompactReport): CompactReport {
+        val gson = Gson()
+        val json = gson.toJson(report)
+        return gson.fromJson(json, CompactReport::class.java)
+    }
```

Comment:
This code is extremely suspicious :sweat_smile: 

I'd really suggest looking into a different way to achieve that without serializing and deserializing the object.